runner/utils/driver/docker/docker.go

```diff
@@ -0,0 +1,255 @@
+package docker
+
+import (
+	"context"
+	"fmt"
+	"io"
+	"os"
+	"strconv"
+	"strings"
+
+	"github.com/docker/docker/api/types"
+	"github.com/docker/docker/api/types/container"
+	"github.com/docker/docker/api/types/filters"
+	"github.com/docker/docker/client"
+	"github.com/docker/go-connections/nat"
+	"github.com/sirupsen/logrus"
+	"github.com/txn2/txeh"
+
+	"github.com/spaceuptech/space-cloud/runner/model"
+	"github.com/spaceuptech/space-cloud/runner/utils/auth"
+)
+
+type docker struct {
+	client *client.Client
+	auth   *auth.Module
+}
+
+func NewDockerDriver(auth *auth.Module) (*docker, error) {
+	cli, err := client.NewClientWithOpts(client.FromEnv, client.WithAPIVersionNegotiation())
+	if err != nil {
+		logrus.Errorf("error creating docker module instance in docker in docker unable to initialize docker client - %v", err)
+		return nil, err
+	}
+	return &docker{client: cli, auth: auth}, nil
+}
+
+// ApplyService creates containers for specified service
+func (d *docker) ApplyService(ctx context.Context, service *model.Service) error {
+	// remove containers if already exits
+	if err := d.DeleteService(ctx, service.ID, service.ProjectID, service.Version); err != nil {
+		logrus.Errorf("error applying service in docker unable delete existing containers - %v", err)
+		return err
+	}
+
+	// todo check host file overiding concern
+	// client for CRUD operation on host file
+	// default location /etc/hosts
+	hostFile, err := txeh.NewHostsDefault()
+	if err != nil {
+		logrus.Errorf("error applying service in docker unable to load host file with suitable default - %v", err)
+		return err
+	}
+
+	service.Labels["internalRuntime"] = string(service.Runtime)
+
+	for _, task := range service.Tasks {
+		// todo get image
+		// for now images have been created locally but not uploaded on docker hub
+		out, err := d.client.ImagePull(ctx, task.Docker.Image, types.ImagePullOptions{})
+		if err != nil {
+			return err
+		}
+		io.Copy(os.Stdout, out)
+
+		// TODO check docker runtime field in host config
+		// expose ports of docker container as specified in task
+		exposedPorts := map[nat.Port]struct{}{}
+		for _, port := range task.Ports {
+			portString := strconv.Itoa(int(port.Port))
+			portWithProtocol := nat.Port(fmt.Sprintf("%s/%s", portString, port.Protocol))
+			exposedPorts[portWithProtocol] = struct{}{}
+		}
+
+		// set environment variables of docker container
+		task.Env["URL"] = fmt.Sprintf("http://service.artifact.svc.cluster.local:4122")
+		token, err := d.auth.GenerateHS256Token(service.ID, service.ProjectID, service.Version)
+		if err != nil {
+			logrus.Errorf("error applying service in docker unable generate token - %v", err)
+			return err
+		}
+		task.Env["TOKEN"] = token
+		envs := []string{}
+		for envName, envValue := range task.Env {
+			envs = append(envs, fmt.Sprintf("%s=%s", envName, envValue))
+		}
+		//
+		// service.Labels["internalServiceId"] = service.ID
+		// service.Labels["internalProjectId"] = service.ProjectID
```

Comment:
Don't be shy. Go ahead and add this. It would be really nice if you could add all `internal` labels together